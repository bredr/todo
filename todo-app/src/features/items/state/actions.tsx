import { ActionType, createAction } from "typesafe-actions";
import { grpcRequest } from "../../../middleware/grpc";
import { Items, Pagination } from "../../../proto/todo_pb";

export const setItems = createAction("SET_ITEMS", action => {
  return (items: Items.AsObject) => action(items);
});

export const loadingItems = createAction("LOADING_ITEMS");

export const getItems = grpcRequest<Pagination, Items>();

export type Actions =
  | ActionType<typeof setItems>
  | ActionType<typeof loadingItems>
  | ActionType<typeof getItems>;
