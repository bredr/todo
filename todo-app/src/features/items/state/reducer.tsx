import produce from "immer";
import { createReducer } from "typesafe-actions";
import { Item } from "../../../proto/todo_pb";
import {
  Actions,
  cancelEditItem,
  editItem,
  getItems,
  loadingItems,
  setItems
} from "./actions";

interface IEditableItem extends Item.AsObject {
  edit: boolean;
}

export interface IState {
  readonly total: number;
  readonly limit: number;
  readonly offset: number;
  readonly items: IEditableItem[];
  readonly loading: boolean;
}

const initState: IState = {
  items: [],
  limit: 100,
  loading: false,
  offset: 0,
  total: 0
};

export default createReducer<IState, Actions>(initState)
  .handleAction(loadingItems, (state, action) =>
    produce(state, draft => {
      draft.loading = true;
      draft.items = [];
      draft.limit = action.payload.limit;
      draft.offset = action.payload.offset;
    })
  )
  .handleAction(setItems, (state, action) =>
    produce(state, draft => {
      draft.loading = false;
      draft.limit = action.payload.limit;
      draft.offset = action.payload.offset;
      draft.total = action.payload.total;
      draft.items = action.payload.itemsList.map(item => ({
        ...item,
        edit: false
      }));
    })
  )
  .handleAction(editItem, (state, action) =>
    produce(state, draft => {
      draft.items = state.items.map(item =>
        item.id === action.payload ? { ...item, edit: true } : item
      );
    })
  )
  .handleAction(cancelEditItem, (state, action) =>
    produce(state, draft => {
      draft.items = state.items.map(item =>
        item.id === action.payload ? { ...item, edit: false } : item
      );
    })
  );
