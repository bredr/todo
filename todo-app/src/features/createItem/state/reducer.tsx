import produce from "immer";
import { createReducer } from "typesafe-actions";
import { Actions, close, open } from "./actions";

export interface IState {
  readonly isOpen: boolean;
}

const initState: IState = {
  isOpen: false
};

export default createReducer<IState, Actions>(initState)
  .handleAction(open, state =>
    produce(state, draft => {
      draft.isOpen = true;
    })
  )
  .handleAction(close, state =>
    produce(state, draft => {
      draft.isOpen = false;
    })
  );
