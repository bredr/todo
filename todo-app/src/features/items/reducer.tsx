import { Item, Items } from "../../proto/todo_pb";
import produce from "immer";
import { createReducer } from "typesafe-actions";
import { setItems, loadingItems, getItems, Actions } from "./actions";

interface IState {
  readonly total: number;
  readonly limit: number;
  readonly offset: number;
  readonly items: Array<Item.AsObject>;
  readonly loading: boolean;
}

const initState: IState = {
  total: 0,
  limit: 100,
  offset: 0,
  items: [],
  loading: false
}

export default createReducer<IState, Actions>(initState).
  handleAction(getItems, (state, action) => produce(state, (draft) => {
    draft.items = [];
    draft.limit = action.payload.request.getLimit();
    draft.offset = action.payload.request.getOffset();
    draft.total = 0;
  })).
  handleAction(loadingItems, (state, _action) => produce(state, (draft) => {
    draft.loading = true;
  })).
  handleAction(setItems, (state, action) => produce(state, (draft) => {
    draft.loading = false;
    draft.limit = action.payload.limit;
    draft.offset = action.payload.offset;
    draft.total = action.payload.total;
    draft.items = action.payload.itemsList;
  }))
