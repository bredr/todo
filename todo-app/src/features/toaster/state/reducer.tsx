import produce from "immer";
import { createReducer } from "typesafe-actions";
import { Actions, addError, addSuccess, clearMessage } from "./actions";

interface IMessage {
  error: boolean;
  message: string;
}
export interface IState {
  readonly messages: IMessage[];
}

const initState: IState = {
  messages: []
};

export default createReducer<IState, Actions>(initState)
  .handleAction(addError, (state, action) =>
    produce(state, draft => {
      draft.messages = [
        { error: true, message: action.payload },
        ...state.messages
      ];
    })
  )
  .handleAction(addSuccess, (state, action) =>
    produce(state, draft => {
      draft.messages = [
        { error: false, message: action.payload },
        ...state.messages
      ];
    })
  )
  .handleAction(clearMessage, state =>
    produce(state, draft => {
      const [, ...remainder] = state.messages;
      draft.messages = remainder;
    })
  );
