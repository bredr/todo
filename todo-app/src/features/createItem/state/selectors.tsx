import { createSelector } from "reselect";
import { IState } from "./reducer";

export const CREATE_SLICE = "CREATE_SLICE";

const sliceSelector = (state: { [CREATE_SLICE]: IState }) =>
  state[CREATE_SLICE];

const createItemOpen = createSelector(
  sliceSelector,
  state => state.isOpen
);

export default { createItemOpen };
