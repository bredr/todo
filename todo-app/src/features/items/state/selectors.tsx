import { createSelector } from "reselect";
import { IState } from "./reducer";

export const ITEMS_SLICE = "ITEMS_SLICE";

const sliceSelector = (state: { [ITEMS_SLICE]: IState }) => state[ITEMS_SLICE];

const items = createSelector(
  sliceSelector,
  state => state.items
);
const loading = createSelector(
  sliceSelector,
  state => state.loading
);
const hasNext = createSelector(
  sliceSelector,
  state => state.total > 0 && state.total - (state.limit + state.offset) > 0
);
const hasPrevious = createSelector(
  sliceSelector,
  state => state.offset > 0 && state.total > 0
);

export default { items, loading, hasNext, hasPrevious };
