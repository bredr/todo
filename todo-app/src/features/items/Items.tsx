import { Paper } from "@material-ui/core";
import CircularProgress from "@material-ui/core/CircularProgress";
import List from "@material-ui/core/List";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import React from "react";
import { useSelector } from "react-redux";
import { CreateForm } from "../createItem/CreateForm";
import create from "../createItem/state/selectors";
import { Item } from "./item/Item";
import selectors from "./state/selectors";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    progress: {
      margin: theme.spacing(2)
    },
    root: {
      width: "100%",
      backgroundColor: theme.palette.background.paper
    },
    padding: {
      "margin-bottom": theme.spacing(2)
    }
  })
);

export const Items: React.FC = () => {
  const items = useSelector(selectors.items);
  const loading = useSelector(selectors.loading);
  const createForm = useSelector(create.createItemOpen);
  const classes = useStyles();

  return (
    <div>
      {createForm && (
        <div className={classes.padding}>
          <CreateForm />
        </div>
      )}
      {loading && <CircularProgress className={classes.progress} />}
      {!loading && items.length > 0 && (
        <div>
          {items.map(item => (
            <div className={classes.padding}>
              {item.edit ? (
                <CreateForm item={item} />
              ) : (
                <Item {...item} key={item.id} />
              )}
            </div>
          ))}
        </div>
      )}
    </div>
  );
};
