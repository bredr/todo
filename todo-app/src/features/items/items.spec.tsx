import { cleanup, render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import { combineReducers, createStore } from "redux";
import createItemReducer, {
  IState as ICreateState
} from "../createItem/state/reducer";
import { CREATE_SLICE } from "../createItem/state/selectors";
import { Items } from "./Items";
import reducer, { IState } from "./state/reducer";
import { ITEMS_SLICE } from "./state/selectors";
const renderWithRedux = (
  ui: JSX.Element,
  {
    initialState,
    store = createStore(
      combineReducers({
        [ITEMS_SLICE]: reducer,
        [CREATE_SLICE]: createItemReducer
      }),
      initialState
    )
  }: {
    initialState?: { [ITEMS_SLICE]: IState; [CREATE_SLICE]: ICreateState };
    store?: any;
  } = {}
) => {
  return { ...render(<Provider store={store}>{ui}</Provider>) };
};

describe("Items", () => {
  afterEach(() => cleanup());

  it("renders spinning wheel when loading", () => {
    const { container } = renderWithRedux(<Items />, {
      initialState: {
        [ITEMS_SLICE]: {
          total: 0,
          limit: 0,
          offset: 0,
          items: [],
          loading: true
        },
        [CREATE_SLICE]: {
          isOpen: false
        }
      }
    });
    expect(container.firstChild).toMatchSnapshot();
  });
  it("renders list of items when not loading", () => {
    const { container } = renderWithRedux(<Items />, {
      initialState: {
        [ITEMS_SLICE]: {
          total: 0,
          limit: 0,
          offset: 0,
          items: [],
          loading: false
        },
        [CREATE_SLICE]: {
          isOpen: false
        }
      }
    });
    expect(container.firstChild).toMatchSnapshot();
  });
  it("renders create form when isOpen", () => {
    const { container } = renderWithRedux(<Items />, {
      initialState: {
        [ITEMS_SLICE]: {
          total: 0,
          limit: 0,
          offset: 0,
          items: [],
          loading: false
        },
        [CREATE_SLICE]: {
          isOpen: true
        }
      }
    });
    expect(container.firstChild).toMatchSnapshot();
  });
});
