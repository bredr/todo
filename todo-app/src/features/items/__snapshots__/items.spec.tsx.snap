// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Items renders create form when isOpen 1`] = `
<div>
  <div
    class="makeStyles-padding-21"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-24 MuiPaper-rounded"
    >
      <div
        class="MuiCardContent-root"
      >
        <form>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
            >
              <div
                class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
                  data-shrink="false"
                  for="description"
                >
                  Description
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    id="description"
                    name="description"
                    type="text"
                    value=""
                  />
                </div>
                
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
            >
              <div
                class="MuiFormControl-root MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                  for="daystocomplete"
                >
                  Days to complete
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    id="daystocomplete"
                    type="number"
                    value="0"
                  />
                </div>
                
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
            >
              <div
                class="MuiFormControl-root MuiTextField-root"
                format="MMMM do"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
                  data-shrink="true"
                >
                  due
                </label>
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiInput-input"
                    readonly=""
                    type="text"
                    value="October 11th"
                  />
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary Mui-disabled MuiButton-fullWidth Mui-disabled"
                disabled=""
                tabindex="-1"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Submit
                </span>
              </button>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
            >
              <button
                aria-label="close"
                class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-fullWidth"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiButton-label"
                >
                  Cancel
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`Items renders list of items when not loading 1`] = `<div />`;

exports[`Items renders spinning wheel when loading 1`] = `
<div>
  <div
    class="MuiCircularProgress-root makeStyles-progress-1 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</div>
`;
